docs/design/proposals/rbd-volume-healer.md

```diff
@@ -0,0 +1,102 @@
+  - [Volume Healer](#volume-healer)
+    - [More thoughts](#more-thoughts)
+
+## Rbd nbd
+
+The rbd CSI plugin will provision new rbd images and attach and mount those
+to workloads. Currently, the default mounter is krbd, which uses the kernel
+rbd driver to mount the rbd images onto the application pod. Here on
+at Ceph-CSI we will also have a userspace way of mounting the rbd images,
+via rbd-nbd.
+
+[Rbd-nbd](https://docs.ceph.com/en/latest/man/8/rbd-nbd/) is a client for
+RADOS block device (rbd) images like the existing rbd kernel module. It
+will map an rbd image to an nbd (Network Block Device) device, allowing
+access to it as a regular local block device.
+
+![csi-rbd-nbd](./images/csi-rbd-nbd.svg)
+
+It’s worth making a note that the rbd-nbd processes will run on the
+client-side, which is inside the `csi-rbdplugin` node plugin.
+
+### Advantages of userspace mounters
+
+- It is easier to add features to rbd-nbd as it is released regularly with
+  Ceph, and more difficult and time consuming to add features to the kernel
+  rbd module as that is part of the Linux kernel release schedule.
+- Container upgrades will be independent of the host node, which means if
+  there are any new features with rbd-nbd, we don’t have to reboot the node
+  as the changes will be shipped inside the container.
+- Because the container upgrades are host node independent, we will be a
+  better citizen in K8s by switching to the userspace model.
+- Unlike krbd, rbd-nbd uses librbd user-space library that gets most of the
+  development focus, and hence rbd-nbd will be feature-rich.
+- Unlike rbd-nbd, krbd exhibits additional drawback, being entirely kernel
+  space impacts fault-tolerance as any kernel panic affects a whole node
+  not only a single pod that is using rbd storage. Thanks to the rbd-nbd’s
+  userspace design, this doesn’t bother here. One might also argue that
+  rbd-nbd has to work with NBD module closely, yes! but remember it's
+  mostly unchanged much from years and consider it to be much stable now.
+  NBD is a generic driver and there will be many eyes on it compared to the
+  rbd driver. rbd driver is vendor-specific logic that might also need
+  changes every now and then when a new feature update is available at Ceph.
+
+### Side effects of userspace mounters
+
+Since the rbd-nbd processes run per volume map on the client side i.e.
+inside the `csi-rbdplugin` node plugin, a restart of the node plugin will
+terminate all the rbd-nbd processes, and there is no way to restore
+these processes back to life currently, which could lead to IO errors
+on all the application pods.
+
+![csi-plugin-restart](./images/csi-plugin-restart.svg)
+
+This is where the Volume healer could help.
+
+## Volume healer
+
+Volume healer runs on the start of rbd node plugin and runs within the
+node plugin driver context.
+
+Volume healer does the below,
+
+- Get the Volume attachment list for the current node where it is running
```

Comment:
Why Ceph RBD uses attacher at all? It does not provide 3rd party attach.

Reply:
@jsafrane True! and we really want to get rid of the attacher and that was one of my concern as raised here  ( Point 2) https://github.com/ceph/ceph-csi/pull/2108#issuecomment-875734208